### ISSUE TITLE
Revert JSONB::toString to produce JSONB::data in order to log actual bind value in LoggerListener and elsewhere

### ISSUE THREAD

**Issue by obourgain** [jOOQ/jOOQ#15751]
### Expected behavior

LoggerListener should log the value actually bound to the statement.

### Actual behavior

LoggerListener prints the query that was executed with the placeholders, and also a version of the query with bound values inlined in the query.
```
        09:58:58,224 DEBUG [org.jooq.tools.LoggerListener                     ] - Executing query          : insert into "mcve"."test" ("cd") values (cast(? as jsonb))
        09:58:58,231 DEBUG [org.jooq.tools.LoggerListener                     ] - -> with bind values      : insert into "mcve"."test" ("cd") values (cast('{"a":20.0}' as jsonb))
```

The JSONB value is used (and that was inserted in the database) is the following.
```
{"a": 20.00}
```

Notice that my value have two zeros in the decimal part while the logged one has only one, also the whitespace after the `:` was removed.

This is because LoggerListener is rendering the query with bound parameters by rendering the query with parameters inline, and with JSONB, the renderer uses toString().

The code is [there](https://github.com/jOOQ/jOOQ/blob/main/jOOQ/src/main/java/org/jooq/tools/LoggerListener.java#L98C1-L98C86)

### Steps to reproduce the problem

* given a table with a JSONB column
* use the LoggerListener with debug log level
* execute an insert with a JSONB value that is not under normalized form
* check the logs

MVCE [https://github.com/obourgain/jOOQ-mcve/tree/wrong_jsonb_value_in_logs](https://github.com/obourgain/jOOQ-mcve/tree/wrong_jsonb_value_in_logs)

### jOOQ Version

3.18.7

### Database product and version

PostgreSQL 15.2 (Debian 15.2-1.pgdg110+1) on aarch64-unknown-linux-gnu, compiled by gcc (Debian 10.2.1-6) 10.2.1 20210110, 64-bit

**Comment by lukaseder** [comment]
Thanks for your message. The real problem here is not with logging, but with the fact that our internal `JSONB` normalisation treats JSON `number` types as `java.lang.Double`, see:

- https://github.com/jOOQ/jOOQ/issues/12173

This shouldn't be a problem in terms of precision with PostgreSQL, as the 2 values are equal, though:

- https://github.com/jOOQ/jOOQ/issues/12173#issuecomment-879701676

However, some improvements may be worth pursuing here, see:

- https://github.com/jOOQ/jOOQ/issues/13547

I'm rejecting this issue here, because:

- Related problems have already been tracked
- The value bound is the one that is being logged. As such, the behaviour is consistent

**Comment by obourgain** [comment]
The value that is bound is not the one logged.
As you can see in the [MCVE](https://github.com/jOOQ/jOOQ-mcve/commit/dc5ccd4261de47331caeb32355b0b23392923be4):
* my value is `{"a": 20.00}` with two trailing 0s
* the logs prints a single trailing 0
```
        09:58:58,231 DEBUG [org.jooq.tools.LoggerListener                     ] - -> with bind values      : insert into "mcve"."test" ("cd") values (cast('{"a":20.0}' as jsonb))
```
* the database stored `{"a": 20.00}` with two trailing 0s

Maybe I made a mistake in my investigation, but I used a debugguer to see where `JSONB.toString()` and `JSONB.data()` where called, and the statement actually used data().

Also, the values may be equals for Postgres, but its documentation explicitly states
```
However, jsonb will preserve trailing fractional zeroes, as seen in this example, even though those are semantically insignificant for purposes such as equality checks.
```
at the end of [this section](https://www.postgresql.org/docs/current/datatype-json.html#JSON-KEYS-ELEMENTS)

**Comment by lukaseder** [comment]
> The value that is bound is not the one logged.

Yes, you're right.

**Comment by lukaseder** [comment]
Before fixing https://github.com/jOOQ/jOOQ/issues/15752, this:

```java
ctx.fetchValue(val(JSONB.jsonb("[2.00]")));
```

Logs:

```sql
select cast('[2.0]' as jsonb)
```

After fixing it, it logs:

```sql
select cast('[2.00]' as jsonb)
```

So, the part you reported is already fixed with #15752. However, I wonder what to do with the fetched result log:

```
Fetched result           : +-----+
                         : |[2.0]|
                         : +-----+
                         : |[2.0]|
                         : +-----+
```

Could be misleading as well, I suppose?

**Comment by obourgain** [comment]
tl;dr Yes, I think the logs must be precise.

long version:
I was doing a update of Jooq 3.14 to 3.18 and the JSONB.toString() doing the normalization made it harder to investigate.
I also understand why it is doing that, and there is probably no perfect solution.
Having spent time investigating this issue, I would prefer JSONB.toString() to not do the normalization. That would avoid the issues I raised and make the debugging experience better.
The trade off is that two equals JSONB instances may have a different textual representation. It may be surprising, but as I mentionned, Postgres already made this choice.

**Comment by lukaseder** [comment]
Most regressions (including performance problems) that had been introduced due to https://github.com/jOOQ/jOOQ/issues/12023 were due to `toString()` having been included in the change, not due to `equals()` and `hashCode()`.

While it would be possible to always make a special case for the `JSONB` type to call its `data()` method instead of `toString()` (including in the `LoggerListener`'s rendering of the `Result`), I do agree, there's not too much benefit of keeping `JSONB.toString()` formatted according to the normalised version of the JSON value.

After all, there isn't even any guarantee that two consecutive parsings of otherwise equal contents produce the same string. jOOQ's internal `JSONObject` representation is based on `HashMap`, which doesn't guarantee iteration order, so while `equals()` and `hashCode()` do not depend on this iteration order, `toString()` still does.

Nevertheless, I feel this change might yet again introduce subtle changes of behaviour that are unexpected for some users, so I won't backport this change.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.19.0

**Comment by obourgain** [comment]
I agree with both the choice of keeping toString() simple, and not backporting to avoid more issues.
Thank you for fixing this.

**Comment by lukaseder** [comment]
I'm reviewing this once more. I certainly agree that the logging should be fixed, but I'm not sure about `toString()`.

The original problem was that precision isn't maintained, and that seems to be an obvious bug. But if we remove normalisation from `toString()`, then all irrelevant whitespace is maintained as well. Users may have grown used to this whitespace being stripped in `toString()`.

**Comment by obourgain** [comment]
I would guess that the main risk with removing whitespaces is breaking tests when the test compares Strings, and indeed that can be annoying for some teams.

**Comment by lukaseder** [comment]
I cannot make up my mind here with respect to the `toString()` implementation, and I don't want to do any rushed changes. After all, this was because of misleading logging, but it's just logging, and doesn't affect any actual behaviour (like previous bugs in this area, where the wrong data was stored). We also abbreviate strings and encode binary data in logs, which could be seen as "misleading," so there too, I don't see anything justifying rushing this.

I do agree that the status quo isn't nice, this is just a matter of prioritisation.

**Comment by lukaseder** [comment]
While I'm still undecided about reverting the `toString()` behaviour, clearly, we should be able to prevent internal calls to the method (which users cannot avoid):

- https://github.com/jOOQ/jOOQ/issues/17578
